src/server/svr_jobfunc.c

```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
+			&pjob->ji_wattr[(int)JOB_ATR_resource], pdef);
+		if (presc &&
+		    ((presc->rs_value.at_flags & ATR_VFLAG_SET) != 0)) {
+			nodect = presc->rs_value.at_val.at_long;
+		}
+	}
+
+	schedselect = expand_select_spec(pjob->ji_wattr[\
+			JOB_ATR_SchedSelect].at_val.at_str);
+
+	if (schedselect == NULL) {
+		LOG_ERR_BUF(err_msg, err_msg_sz-1,
+			"schedselect expand error", -1, id)
+		goto recreate_exec_vnode_exit;
+	}
+
+	new_exec_vnode[0] = '\0';
+	new_exec_host[0] = '\0';
+	new_exec_host2[0] = '\0';
+	deallocated_execvnode[0] = '\0';
+
+	prdefvntype = find_resc_def(svr_resc_def, "vntype", svr_resc_size);
+	prdefarch = find_resc_def(svr_resc_def, "arch", svr_resc_size);
+	/* There's a 1:1:1 mapping among exec_vnode parenthesized entries, exec_host, */
+	/* and exec_host2,  */
+	entry0 = 0;	/* exec_vnode_deallocated entries */
+	entry = 0;	/* exec_vnode entries */
+	h_entry = 0;	/* exec_host* entries */
+	sel_entry = 0;	/* select and schedselect entries */
+	f_entry = 0;	/* number of freed sister nodes */
+	paren = 0;
+	for (	chunk = parse_plus_spec_r(exec_vnode, &last, &hasprn),
+	     	chunk1 = parse_plus_spec_r(exec_host, &last1, &hasprn1),
+	     	chunk2 = parse_plus_spec_r(exec_host2,&last2, &hasprn2),
+	     	chunk3 = parse_plus_spec_r(schedselect, &last3, &hasprn3);
+		(chunk != NULL) && (chunk1 != NULL) && (chunk2 != NULL) && (chunk3 != NULL);
+		chunk = parse_plus_spec_r(last, &last, &hasprn) ) {
+
+		paren += hasprn;
+		strncpy(chunk_buf, chunk, chunk_buf_sz-1);
+		if (parse_node_resc(chunk, &noden, &nelem, &pkvp) == 0) {
+
+			/* see if previous entry already matches this */
+			if ((pnode == NULL) || 
+				(strcmp(pnode->nd_name, noden) != 0)) {
+				pnode = find_nodebyname(noden);
+			}
+
+			if (pnode == NULL) { /* should not happen */
+
+				LOG_EVENT_BUF_ARG1(err_msg,err_msg_sz-1,
+					"no node entry for %s", noden,
+					pjob->ji_qs.ji_jobid)
+				goto recreate_exec_vnode_exit;
+			}
+
+			if ((strcmp(pnode->nd_moms[0]->mi_host,
+						ms_fullhost) == 0) &&
+			     (pnode->nd_moms[0]->mi_port == ms_port) &&
+			     (vnodelist != NULL) &&	
+			      in_string_list(noden, '+', vnodelist)) {
+				LOG_EVENT_BUF_ARG1(err_msg,err_msg_sz-1,
+				 "Can't free '%s' since it's on an "
+				"MS host", noden, pjob->ji_qs.ji_jobid);
+				goto recreate_exec_vnode_exit;
+			}
+
+			if ( (vnodelist != NULL) &&	
+			      in_string_list(noden, '+', vnodelist) &&
+				(pnode->nd_attr[ND_ATR_ResourceAvail].at_flags & ATR_VFLAG_SET) != 0) {
+				prs = (resource *)GET_NEXT(pnode->nd_attr[ND_ATR_ResourceAvail].at_val.at_list);
+				while (prs) {
+					if ( (prdefvntype != NULL) &&
+						(prs->rs_defin == prdefvntype) &&
+						(prs->rs_value.at_flags & ATR_VFLAG_SET) != 0 ) {
+						struct array_strings *as;
+						int	l;
+						as = prs->rs_value.at_val.at_arst;
+						for (l = 0; l < as->as_usedptr; l++) {
+							if ((strcmp(as->as_string[l], CRAY_LOGIN) == 0) ||
+								(strcmp(as->as_string[l], CRAY_COMPUTE) == 0)) {
```

Comment:
Thanks @bhroam for bringing this to my attention.  I agree with the reasoning behind your suggestion.  But I fear it may be a bit complex.  I have asked a few questions to @bayucan in the forum related to the external design for this feature.  If it is expected that linux systems managed by the same PBS that is managing a Cray X* series system, then @bhroam code suggestion of strstr for "cray" could prevent those linux systems from being ramped down (see 3a of my forum comment).  Of course this "cray" keyword could be called out in the external design, so customers that want to use node ramp down will know what NOT to set for vntype.  This may be a good idea anyway since vntype may be set by admins.